### DISCUSSION TITLE
Extending Saleor with a B2B module

### DISCUSSION THREAD

**Discussion by Dumi23** [saleor/saleor#13220] [Show and tell]
Starting this year I have worked on an eCommerce project using saleor, the scope and the requirements of the project included some necessary b2b functionalities that saleor did not have, but on the other side, the time constraints lead to some quite messy implementations of those same functionalities.

Coming to the conclusion that saleor is indeed a stable platform that I would like to continue to work with I am opening a fork with these functionalities in mind and make them more in line with how the code works with the original saleor

**PROPOSED CHANGES**
Introduction of new models: 

- `CompanyInfo`
- `CustomerGroup`
- `CategoryDiscount`

Adding a foreign key of `CompanyInfo` to model `account.User`

With `CompanyInfo` model customers can request access to the b2b portion of the shop and be approved to be added to groups (model `CustomerGroup`) and have access to special `CategoryDiscounts` (at a later stage we plan on introducing both `CollectionDiscounts` and  `ProductDiscounts`)

By the time anyone is reading this the first initial commit will be pushed to the fork with the underlying mutations and types done 
the next step would be the solve the problem of automatically applying discounts for the customers that are in `CustomerGroup` and their proper calculation.

I would like to ask the Saleor Team if it is ok to open a pull request for this

Link to the fork: [https://github.com/Dumi23/saleor/tree/saleorb2b](url)
If anyone has any questions they can freely contact me on Discord with my username being: .dumi
Or by mail: [okiljdamjan@gmail.com](url)

**Comment by maarcingebala** [comment]
Hey @Dumi23, thanks for opening this discussion. We're working on similar features right now; please have a look at these two RFCs:
- https://github.com/saleor/saleor/issues/13070 - in this RFC we propose introducing customer groups, just as in your case, but more generic
- https://github.com/saleor/saleor/issues/12794 - this RFC is already in progress and will introduce a new generic promotions engine, where category discounts are just one of many available predicates for creating discounts. The new promotions engine will also support assigning discounts to customer groups.

Regarding granting access to B2B, I think we would like to solve it by creating a dedicated B2B channel and assigning access to it via customer groups. Any additional info might be stored in metadata, so there is no need for an extra model here.

If you would like to help as collaborating on these changes let us know :)

**Reply by Dumi23** [reply]
I would love to be a contributor I opened a pull request for these proposed changes you can check it out if you have time and maybe I can get some feedback on them.

And thanks for the response.
Have a great day

**Reply by maarcingebala** [reply]
The code looks solid but to introduce such significant changes, we first need to discuss the shape of the code and API in an RFC. We have two mentioned RFCs already that seem to cover similar cases. Maybe instead of working on the PR, you could let us know if the changes we're planning would solve your use cases. If not, we could think about how to improve the RFCs.

**Reply by Dumi23** [reply]
Thank you for the great feedback. 
PR in my mind was never intended for merging since the RFCs somewhere differ on in their specs from my own intentions but since it was my first PR to a big open source  codebase like this one I wanted to see the feedback.

I will take an example from `CustomerGroups` in the RFC it is stated that customers can belong to many groups while my use case limits customers to belong to just one group.

But nevertheless I will keep in mind your suggestion.
Thanks for your response and feedback it is much appreciated.
Have a great day.

**Reply by maarcingebala** [reply]
You're welcome!
